src/HotChocolate/Raven/src/Data/Pagination/RavenPagingContainer.cs

```diff
@@ -0,0 +1,182 @@
+            //    concurrent call on count async waits, until we are done with the query
+            if (_totalCount is null)
+            {
+                totalCountCompletionSource = new TaskCompletionSource<int>();
+
+                // capture the source so it is not in the outer scope
+                var source = totalCountCompletionSource;
+                var originalTotalCount =
+                    Interlocked.CompareExchange(ref _totalCount, new TaskCache(source.Task), null);
+
+                // in case CountAsync was already executed we reset the completion source and await
+                // the CountAsync call so that we do not have concurrent requests on the async
+                // session
+                if (originalTotalCount is not null)
+                {
+                    totalCountCompletionSource = null;
+                    await originalTotalCount.Execute();
+                }
+            }
+
+            // We only load the query stats when we not already have fetched them.
+            IAsyncEnumerator<StreamResult<TEntity>> cursor;
+            if (totalCountCompletionSource is { })
+            {
+                cursor = await _query
+                    .AsAsyncEnumerable(out var stats, cancellationToken)
+                    .ConfigureAwait(false);
+
+                totalCountCompletionSource?.SetResult(stats.TotalResults);
+            }
+            else
+            {
+                cursor = await _query
+                    .AsAsyncEnumerable(cancellationToken)
+                    .ConfigureAwait(false);
+            }
+
+            // page through the response and create the array
+            var list = new List<IndexEdge<TEntity>>();
+            var index = offset;
+            while (await cursor.MoveNextAsync().ConfigureAwait(false))
+            {
+                list.Add(IndexEdge<TEntity>.Create(cursor.Current.Document, index++));
+            }
+
+            return list;
+        }
+        catch (OperationCanceledException)
+        {
+            totalCountCompletionSource?.SetCanceled(cancellationToken);
+            throw;
+        }
+        catch (Exception ex)
+        {
+            totalCountCompletionSource?.SetException(ex);
+            throw;
+        }
+        finally
+        {
+            if (totalCountCompletionSource is { Task.IsCompleted: false })
+            {
+                totalCountCompletionSource.SetCanceled(cancellationToken);
+            }
+        }
+    }
+
+    public async ValueTask<List<TEntity>> ToListAsync(CancellationToken cancellationToken)
+    {
+        return await _query.ToListAsync(cancellationToken);
+    }
+
+    public RavenPagingContainer<TEntity> Skip(int skip)
+    {
+        _query = _query.Skip(skip);
+        return this;
+    }
+
+    public RavenPagingContainer<TEntity> Take(int take)
+    {
+        _query = _query.Take(take);
+        return this;
+    }
+
+    private sealed class TaskCache
```

Comment:
this is not a task cache ... more something that makes sure you only resolve something once. Task cache for dataloader for instance in fact caches tasks.

Reply:
it looks more like a Lazy

Reply:
TaskHolder